### PR TITLE
[Geometry] Add method intersectionWithEdge in Edge structure

### PR DESCRIPTION
**Pull request by epernod**
based on #4028 

Add method to detect intersection between 2 edges in 2D and 3D.
Add tests with tricky cases
Will update version in EdgeSetTopologyContainer which is not really working later.

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: Sofa/framework/Geometry/src/sofa/geometry/Edge.h [hunk 1]
```diff
@@ -159,6 +204,109 @@ struct Edge
...
+                intersection = sofa::type::Vec<2, T>(sofa::InvalidID, sofa::InvalidID);
+                return false;
+            }
+            else
+            {
+                intersection = pA + alpha * AB;
+                return true;
+            }
+        }
+        else
```
Comment (fredroy) [inline Sofa/framework/Geometry/src/sofa/geometry/Edge.h:262]: Same remark as https://github.com/sofa-framework/sofa/pull/4028

**Review by fredroy — CHANGES_REQUESTED** [review]

File: Sofa/framework/Geometry/src/sofa/geometry/Edge.h [hunk 2]
```diff
@@ -159,6 +204,109 @@ struct Edge
...
+            // (pX - pY).dot(pD - pC) = 0
+            
+            // We need to find alpha and beta that suits: 
+            // [ (pA - pC) + alpha(pB - pA) - beta(pD - pC) ].dot(pB - pA) = 0
+            // [ (pA - pC) + alpha(pB - pA) - beta(pD - pC) ].dot(pD - pC) = 0
+            const auto CA = pA - pC;
+
+            // Writting d[CA/AB] == (pA - pC).dot(pB - pA) and subtituting beta we obtain:
+            // beta = (d[CA/CD] + alpha * d[AB/CD]) / d[CD/CD]
+            // alpha = ( d[CA/CD]*d[CD/AB] - d[CA/AB]*d[CD/CD] ) / ( d[AB/AB]*d[CD/CD] - d[AB/CD]*d[AB/CD])
```
Comment (bakpaul) [inline Sofa/framework/Geometry/src/sofa/geometry/Edge.h:276]: I tried to do the math on my side and it's good !

File: Sofa/framework/Geometry/src/sofa/geometry/Edge.h [hunk 3]
```diff
@@ -159,6 +204,109 @@ struct Edge
...
+        // pX = pA + alpha (pB - pA)
+        // pY = pC + beta (pD - pC)
+        const auto AB = pB - pA;
+        const auto CD = pD - pC;
+
+        if constexpr (std::is_same_v < Node, sofa::type::Vec<2, T> >)
+        {
+            // in 2D we have 2 segment equations and 2 unknowns so direct solving of pX = pY is possible
+            // pA + alpha (pB - pA) = pC + beta (pD - pC)
+            // alpha = ((Cy - Ay)(Dx - Cx) - (Cx - Ax)(Dy - Cy)) / ((By - Ay)(Dx - Cx) - (Bx - Ax)(Dy - Cy))
```
Comment (bakpaul) [inline Sofa/framework/Geometry/src/sofa/geometry/Edge.h:238]: Math are ok for me

**Review by bakpaul — APPROVED** [review]

File: Sofa/framework/Geometry/src/sofa/geometry/Edge.h (hunk 2 above)
Comment (epernod) [reply Sofa/framework/Geometry/src/sofa/geometry/Edge.h:276]: ouf! 😄

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

File: Sofa/framework/Geometry/src/sofa/geometry/Edge.h [hunk 4]
```diff
@@ -207,6 +207,109 @@ struct Edge
...
+                return false;
+            }
+
+            const T alpha = alphaNom / alphaDenom;
+            const T beta = (dCACD + alpha * dABCD) / dCDCD;
+
+            const Node pX = pA + alpha * AB;
+            const Node pY = pC + beta * CD;
+
+            if (alpha < 0 || beta < 0 || (pY - pX).norm2() > EQUALITY_THRESHOLD )
```
Comment (epernod) [inline Sofa/framework/Geometry/src/sofa/geometry/Edge.h]: alpha > 1 and beta >1 should also be checked. this means the intersection is out of the edge.

**Pull request merged**